app/src/main/java/org/kiwix/kiwixmobile/local_file_transfer/LocalFileTransferActivity.kt

```diff
@@ -0,0 +1,414 @@
+ *
+ * The module is used for transferring ZIM files from one device to another, from within the
+ * app. Two devices are connected to each other using WiFi Direct, followed by file transfer.
+ *
+ * File transfer involves two phases:
+ * 1) Handshake with the selected peer device, using [PeerGroupHandshakeAsyncTask]
+ * 2) After handshake, starting the files transfer using [SenderDeviceAsyncTask] on the sender
+ * device and [ReceiverDeviceAsyncTask] files receiving device
+ */
+@SuppressLint("GoogleAppIndexingApiWarning")
+class LocalFileTransferActivity : BaseActivity(),
+  WifiDirectManager.Callbacks {
+  @Inject
+  var alertDialogShower: AlertDialogShower? = null
+
+  @Inject
+  var wifiDirectManager: WifiDirectManager? = null
+
+  @Inject
+  var locationManager: LocationManager? = null
+
+  @BindView(R.id.toolbar)
+  var actionBar: Toolbar? = null
+
+  @BindView(R.id.text_view_device_name)
+  var deviceName: TextView? = null
+
+  @BindView(R.id.progress_bar_searching_peers)
+  var searchingPeersProgressBar: ProgressBar? = null
+
+  @BindView(R.id.list_peer_devices)
+  var peerDeviceList: RecyclerView? = null
+
+  @BindView(R.id.text_view_empty_peer_list)
+  var textViewPeerDevices: TextView? = null
+
+  @BindView(R.id.recycler_view_transfer_files)
+  var filesRecyclerView: RecyclerView? = null
+  private var isFileSender = false // Whether the device is the file sender or not
+  private var filesForTransfer = ArrayList<FileItem>()
+  private var fileListAdapter: FileListAdapter? = null
+  private var wifiPeerListAdapter: WifiPeerListAdapter? = null
+  override fun onCreate(savedInstanceState: Bundle?) {
+    super.onCreate(savedInstanceState)
+    setContentView(R.layout.activity_local_file_transfer)
+    /*
+     * Presence of file Uris decides whether the device with the activity open is a sender or receiver:
+     * - On the sender device, this activity is started from the app chooser post selection
+     * of files to share in the Library
+     * - On the receiver device, the activity is started directly from within the 'Get Content'
+     * activity, without any file Uris
+     * */
+    val filesIntent = intent
+    val fileUriArrayList: ArrayList<Uri>?
+    fileUriArrayList = filesIntent.getParcelableArrayListExtra(Intent.EXTRA_STREAM)
+    isFileSender = fileUriArrayList != null && fileUriArrayList.size > 0
+    setSupportActionBar(actionBar)
+    actionBar!!.setNavigationIcon(R.drawable.ic_close_white_24dp)
+    actionBar!!.setNavigationOnClickListener { finish() }
+    wifiPeerListAdapter = WifiPeerListAdapter(
+      WifiP2pDelegate { wifiP2pDevice: WifiP2pDevice? ->
+        wifiDirectManager!!.sendToDevice(wifiP2pDevice!!)
+      }
+    )
+    peerDeviceList!!.adapter = wifiPeerListAdapter
+    peerDeviceList!!.layoutManager = LinearLayoutManager(this)
+    peerDeviceList!!.setHasFixedSize(true)
+    if (isFileSender) {
+      for (i in fileUriArrayList.indices) {
```

Comment:
if you see a for loop in kotlin then something has probably gone wrong. What you want here is to use the `map` function to create a new list from fileUriArrayList

Reply:
I'm not quite sure, how to use the `map` function for this purpose. Could you maybe help me out please?

Reply:
The docs on [transformation](https://kotlinlang.org/docs/reference/collection-transformations.html) should help.

The method definition itself is quite illuminating
![image](https://user-images.githubusercontent.com/3358282/89618177-86acac00-d883-11ea-9961-8fa94407b436.png)